File: priv/combined_metrics/samples/scope_and_assignment/used_only_once/good/query_builder.js
// Query/struct builder — GOOD: intermediate results are inlined or chained.

class QueryBuilder {
  buildSearchQuery(filters) {
    return from("products")
      .where({ category: filters.category })
      .whereLte("price", filters.maxPrice)
      .whereGt("stock", 0)
      .orderBy("inserted_at")
      .limit(filters.limit);
  }

  buildUserObject(attrs) {
    return {
      name: attrs.name.trim(),
      email: attrs.email.toLowerCase(),
      role: attrs.role || "guest",
    };
  }

  formatReport(data) {
    const header = `=== ${data.title.toUpperCase()} ===`;
    const body = data.rows.map((r) => this._formatRow(r)).join("\n");
    return `${header}\n${body}`;
  }

  buildNotification(event) {
    return {
      subject: `Event: ${event.name}`,
      to: event.user.email,
      body: renderTemplate(loadTemplate(event.type), event),
    };
  }

  composeUrl(baseUrl, path, queryParams) {
    return `${baseUrl}${path}?${new URLSearchParams(queryParams).toString()}`;
  }

  _formatRow(row) {
    return `${row.label}: ${row.value}`;
  }
}

function loadTemplate(t) {
  return `template_${t}`;
}

function renderTemplate(t, _e) {
  return t;
}

function from(_) {
  const q = {
    where: () => q,
    whereLte: () => q,
    whereGt: () => q,
    orderBy: () => q,
    limit: () => q,
  };
  return q;
}

module.exports = { QueryBuilder };

File: priv/combined_metrics/samples/scope_and_assignment/used_only_once/good/query_builder.rb
# Query/struct builder — GOOD: intermediate results are inlined or chained.

require "uri"

class QueryBuilder
  def build_search_query(filters)
    Query.from("products")
         .where(category: filters[:category])
         .where_lte(:price, filters[:max_price])
         .where_gt(:stock, 0)
         .order_by(:inserted_at)
         .limit(filters[:limit])
  end

  def build_user_hash(attrs)
    {
      name: attrs[:name].strip,
      email: attrs[:email].downcase,
      role: attrs[:role] || :guest
    }
  end

  def format_report(data)
    header = "=== #{data[:title].upcase} ==="
    body = data[:rows].map { |r| format_row(r) }.join("\n")
    "#{header}\n#{body}"
  end

  def build_notification(event)
    {
      subject: "Event: #{event[:name]}",
      to: event[:user][:email],
      body: render_template(load_template(event[:type]), event)
    }
  end

  def compose_url(base_url, path, query_params)
    "#{base_url}#{path}?#{URI.encode_www_form(query_params)}"
  end

  private

  def format_row(row)
    "#{row[:label]}: #{row[:value]}"
  end

  def load_template(type)
    "template_#{type}"
  end

  def render_template(template, _event)
    template
  end
end

class Query
  def self.from(_); new; end
  def where(*); self; end
  def where_lte(*); self; end
  def where_gt(*); self; end
  def order_by(*); self; end
  def limit(*); self; end
end
